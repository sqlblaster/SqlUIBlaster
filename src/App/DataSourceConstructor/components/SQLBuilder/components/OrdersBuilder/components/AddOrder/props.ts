import {
  actionCreators,
  PickActionCreators
} from 'src/store/action-creators';
import { PickStates } from 'src/store/models/State';
import { PopoverManagerProps } from '../../../common/WithPopoverManagement';

export type AddOrderOwnProps = PopoverManagerProps;

export type AddOrderStateProps = PickStates<'orders'>;

export type AddOrderDispatchProps = PickActionCreators<'addOrder'>;

export const AGMapDispatchToProps: AddOrderDispatchProps = {
  addOrder: actionCreators.addOrder
};

export type AddOrderProps = AddOrderDispatchProps & AddOrderStateProps;

export type AddOrderViewProps = AddOrderOwnProps & AddOrderDispatchProps & AddOrderStateProps;
