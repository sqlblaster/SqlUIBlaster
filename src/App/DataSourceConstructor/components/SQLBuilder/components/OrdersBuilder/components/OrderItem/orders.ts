enum SortOrderEnum {
  'ascending',
  'descending'
}

export type SortOrder = keyof typeof SortOrderEnum;

export const sortOrders: SortOrder[] = [
  'ascending',
  'descending'
];
