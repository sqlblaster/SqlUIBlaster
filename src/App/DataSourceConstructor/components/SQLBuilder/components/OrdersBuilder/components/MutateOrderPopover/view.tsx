import * as React from 'react';
import { ColumnSelector } from '../../../common/ColumnSelector';
import { PopoverPaper } from '../../../common/PopoverPaper';
import { MutateOrderPopoverViewProps } from './props';

export const MutateOrderPopover: React.FC<
  MutateOrderPopoverViewProps
> = props => {
  const {
    popoverPositionUpdater,
    onColumnSelected,
    canShowColumnBranch
  } = props;

  return (
    <PopoverPaper {...props}>
      <div className='mutate-order'>
        <ColumnSelector
          popoverPositionUpdater={popoverPositionUpdater}
          onColumnSelected={onColumnSelected}
          canShowColumnBranch={canShowColumnBranch}
        />
      </div>
    </PopoverPaper>
  );
};
