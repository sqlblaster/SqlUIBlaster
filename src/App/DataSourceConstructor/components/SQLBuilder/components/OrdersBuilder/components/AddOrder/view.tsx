import * as React from 'react';
import { AddEntity } from '../../../common/AddEntity';
import { MutateOrderPopover } from '../MutateOrderPopover';
import { AddOrderViewProps } from './props';

export const addOrderText = 'Pick a field to sort by';

// FIXME: make padding clickable too
export class AddOrder extends React.Component<AddOrderViewProps, any> {
  private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

  constructor(props: AddOrderViewProps) {
    super(props);

    this.popoverLauncher = React.createRef<HTMLDivElement>();
  }

  public render() {
    const {
      orders,
      addOrder,
      isOpen,
      handleClose,
      handleOpen
    } = this.props;

    return (
      <>
        <AddEntity
          onClick={handleOpen}
          reference={this.popoverLauncher}
          showText={!(orders && orders.length)}
        >
          {addOrderText}
        </AddEntity>
        { isOpen && (// TODO: resolve this problem another way
          <MutateOrderPopover
            action={addOrder}
            isOpen={isOpen}
            anchorEl={this.popoverLauncher.current}
            onClose={handleClose}
          />
        ) }
      </>
    );
  }
}
