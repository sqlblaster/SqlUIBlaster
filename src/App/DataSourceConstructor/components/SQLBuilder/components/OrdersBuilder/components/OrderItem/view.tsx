import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { RemoveButton } from 'src/components/query-builder/App/DataSourceConstructor/components/SQLBuilder/components/common/Icons/RemoveButton';
import { SequentialColumnView } from '../../../common/SequentialColumnView';
import { CheckedSelect } from '../../../FilterBuilder/components/MutateFilterPopover/SwitchMutator/CheckedSelect';
import { MutateOrderPopover } from '../MutateOrderPopover';
import { SortOrder, sortOrders } from './orders';
import { OrderItemViewProps } from './props';
import { orderItemStyles } from './styles';

export const OrderItem = withStyles(orderItemStyles)(
  class extends React.Component<OrderItemViewProps> {
    private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

    constructor(props: OrderItemViewProps) {
      super(props);

      this.popoverLauncher = React.createRef<HTMLDivElement>();
    }

    public render() {
      const {
        order,
        updateOrder,
        classes,
        isOpen,
        handleOpen,
        handleClose,
        handleOrderRemoval,
        handleSortOrderChange
      } = this.props;

      const { id, column, sortOrder } = order;

      return (
        <>
          <div className={classes.root}>
            <div
              ref={this.popoverLauncher}
              onClick={handleOpen}
              className={classes['column-view']}
            >
              <SequentialColumnView {...column} />
            </div>
            <CheckedSelect<SortOrder>
              items={sortOrders}
              onSelect={handleSortOrderChange}
              selectedItem={sortOrder}
              hasBorder={false}
            />
            <RemoveButton onClick={handleOrderRemoval(id)} />
          </div>
          { isOpen && (
            <MutateOrderPopover
              isOpen={isOpen}
              anchorEl={this.popoverLauncher.current}
              action={updateOrder}
              onClose={handleClose}
              order={order}
            />
          ) }
        </>
      );
    }
  }
);
