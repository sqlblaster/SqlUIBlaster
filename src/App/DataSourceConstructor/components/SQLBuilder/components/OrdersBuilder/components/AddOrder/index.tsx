import * as React from 'react';
import { connect } from 'react-redux';
import { State } from 'src/store/models/State';
import { WithPopoverManagement } from '../../../common/WithPopoverManagement';
import {
  AddOrderDispatchProps,
  AddOrderProps,
  AddOrderStateProps,
  AGMapDispatchToProps
} from './props';
import { AddOrder as View } from './view';

export const AddOrder = connect<AddOrderStateProps, AddOrderDispatchProps, {}, State>(
  ({ orders }) => ({ orders }),
  AGMapDispatchToProps
)(
  class extends React.Component<AddOrderProps> {
    public static ViewWithPopoverManagement = WithPopoverManagement(View);

    public render() {
      return <AddOrder.ViewWithPopoverManagement addOrder={this.props.addOrder} {...this.props} />;
    }
  }
);
