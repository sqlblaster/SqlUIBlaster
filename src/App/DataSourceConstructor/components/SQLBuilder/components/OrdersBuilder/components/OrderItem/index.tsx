import * as React from 'react';
import { connect } from 'react-redux';
import { State } from 'src/store/models/State';
import { WithPopoverManagement } from '../../../common/WithPopoverManagement';
import { Order } from './model';
import {
  OIMapDispatchToProps,
  OrderItemDispatchProps,
  OrderItemOwnProps,
  OrderItemProps,
  OrderItemViewProps as ViewProps
} from './props';
import { OrderItem as View } from './view';

export const OrderItem = connect<
  {},
  OrderItemDispatchProps,
  OrderItemOwnProps,
  State
>(
  null,
  OIMapDispatchToProps
)(
  class extends React.Component<OrderItemProps> {
    public static ViewWithPopoverManagement = WithPopoverManagement(View);

    public handleOrderRemoval: ViewProps['handleOrderRemoval'] = (id) => () =>
      this.props.removeOrder(id)

    public handleSortOrderChange: ViewProps['handleSortOrderChange'] = (sortOrder) => {
      const { updateOrder, order } = this.props;
      updateOrder(
        new Order(order, {
          sortOrder
        })
      );
    }

    public render() {
      return (
        <OrderItem.ViewWithPopoverManagement
          handleOrderRemoval={this.handleOrderRemoval}
          handleSortOrderChange={this.handleSortOrderChange}
          {...this.props}
        />
      );
    }
  }
);
