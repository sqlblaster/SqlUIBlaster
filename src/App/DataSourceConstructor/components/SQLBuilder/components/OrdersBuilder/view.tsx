import { Typography, withStyles } from '@material-ui/core';
import * as React from 'react';
import { AddOrder } from './components/AddOrder';
import { OrderItem } from './components/OrderItem';
import { OrdersBuilderViewProps } from './props';
import { ordersBuilderStyles } from './styles';

export const noColumnsAvailableText = 'No columns available for sorting';

export const OrdersBuilder = withStyles(ordersBuilderStyles)((({
  orders,
  aggregations,
  groupings,
  classes
}) => (
  <>
    {!(aggregations.length && !groupings.length) ? (
      <div className={classes.root}>
        {orders && orders.length ? (
          <div className={classes.orders}>
            {orders.map(order => (
              <OrderItem key={order.id} order={order} />
            ))}
          </div>
        ) : null}
        <AddOrder />
      </div>
    ) : (
      <Typography className={classes['no-column']}>
        {noColumnsAvailableText}
      </Typography>
    )}
  </>
)) as React.FC<OrdersBuilderViewProps>);
