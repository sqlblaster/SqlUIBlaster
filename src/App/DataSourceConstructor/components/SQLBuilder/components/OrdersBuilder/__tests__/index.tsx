import {
  fireEvent,
  getAllByText as getAllByTextOn,
  getByText as getByTextOn,
  queryByText as queryByTextOn
} from 'react-testing-library';
import { regex } from 'src/components/query-builder/App/DataSourceConstructor/test-utils/utils';
import { getGeneratedQuery, render } from '../../../__tests__/render.utils';
import { ordersBuilderLabel } from '../../../labels';
import { selectAColumn } from '../../common/ColumnSelector/__tests__/utils/select-a-column';
import {
  clickAddAGrouping,
  clickAddMoreGrouping
} from '../../GroupingsBuilder/components/AddGrouping/__tests__/utils/click-add-a-grouping';
import {
  clickAddAnOrder,
  clickAddMoreOrder
} from '../../OrdersBuilder/components/AddOrder/__tests__/utils/click-add-an-order';
import { selectATable } from '../../TableSelector/__tests__/utils/select-table';
import { addOrderText } from '../components/AddOrder/view';
import { sortOrders } from '../components/OrderItem/orders';

describe('Orders Builder', () => {
  const { getByText, queryByText, getByTestId } = render();

  const columns = selectATable();

  const ordersBuilderPanel = getByTextOn(
    document.body,
    regex(ordersBuilderLabel)
  );

  const clickOnOrder = (columnName: string) => {
    fireEvent.click(getByTextOn(ordersBuilderPanel, regex(columnName)));
  };

  const removeAnOrder = () => {
    fireEvent.click(getAllByTextOn(ordersBuilderPanel, /clear/i)[0]);
  };

  const detailsColumn = columns[1];
  const createdAtColumn = columns[2];

  const checkWhiteList = (whiteList: string[]) => {
    const columnSelectorPanel = getByTestId('column-selector');

    columns.forEach(column => {
      if (whiteList.includes(column.name) || column.foreignTableName) {
        expect(
          getByTextOn(columnSelectorPanel, column.name)
        ).toBeInTheDocument();
      } else {
        expect(
          queryByTextOn(columnSelectorPanel, column.name)
        ).not.toBeInTheDocument();
      }
    });
  };

  it('should show the add order placeholder', () => {
    expect(getByText(addOrderText)).toBeInTheDocument();
  });

  it(`should add an order`, () => {
    clickAddAnOrder();

    fireEvent.click(getByText(detailsColumn.name));

    expect(getGeneratedQuery()).toMatchSnapshot();
  });

  it('should hide the add order placeholder', () => {
    expect(queryByText(addOrderText)).not.toBeInTheDocument();
  });

  it('should update order', () => {
    clickOnOrder(detailsColumn.name);

    selectAColumn(createdAtColumn);

    expect(getGeneratedQuery()).toMatchSnapshot();
  });

  it(`should add foreign column order properly
      and hide already added column`, () => {
    clickAddMoreOrder();

    // hide already added column
    const columnSelector = getByTestId('column-selector');
    columns.forEach(column => {
      if (column.name === createdAtColumn.name) {
        expect(
          queryByTextOn(columnSelector, createdAtColumn.name)
        ).not.toBeInTheDocument();
      } else {
        expect(queryByTextOn(columnSelector, column.name)).toBeInTheDocument();
      }
    });

    const productColumn = columns[5];
    const {
      foreignColumns: foreignProductColumns,
      foreignColumnsPanel: foreignProductColumnsPanel
    } = selectAColumn(productColumn, { expand: true });
    const currencyColumn = foreignProductColumns[4];
    const { foreignColumns: foreignCurrencyColumns } = selectAColumn(
      currencyColumn,
      {
        expand: true,
        container: foreignProductColumnsPanel
      }
    );
    const currencyName = foreignCurrencyColumns[1];
    selectAColumn(currencyName);

    expect(getGeneratedQuery()).toMatchSnapshot();
  });

  it(`should remove all orders except the grouping columns
      when a grouping added and show only picked grouping column
      available for order column`, () => {
    clickAddAGrouping();

    selectAColumn(detailsColumn);

    clickAddMoreOrder();

    const columnSelectorPanel = getByTestId('column-selector');

    columns.forEach(column => {
      if (column.name === detailsColumn.name || column.foreignTableName) {
        expect(
          getByTextOn(columnSelectorPanel, column.name)
        ).toBeInTheDocument();
      } else {
        expect(
          queryByTextOn(columnSelectorPanel, column.name)
        ).not.toBeInTheDocument();
      }
    });

    selectAColumn(detailsColumn, { container: columnSelectorPanel });

    const ascendingOrder = sortOrders[0];
    const descendingOrder = sortOrders[1];
    fireEvent.click(getByTextOn(ordersBuilderPanel, regex(ascendingOrder)));
    fireEvent.click(getByTextOn(ordersBuilderPanel, regex(descendingOrder)));

    expect(getGeneratedQuery()).toMatchSnapshot();
  });

  it('should show only not selected columns among grouping columns only', () => {
    removeAnOrder();

    clickAddMoreGrouping();
    selectAColumn(createdAtColumn);

    clickAddAnOrder();

    checkWhiteList([createdAtColumn.name, detailsColumn.name]);

    selectAColumn(detailsColumn, { container: ordersBuilderPanel });

    clickAddMoreOrder();

    checkWhiteList([createdAtColumn.name]);
  });

  it('should show the add order placeholder again', () => {
    removeAnOrder();

    expect(getByText(addOrderText)).toBeInTheDocument();
  });
});
