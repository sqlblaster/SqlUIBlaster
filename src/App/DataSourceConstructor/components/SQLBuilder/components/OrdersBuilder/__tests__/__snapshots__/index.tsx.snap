// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orders Builder should add an order 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
ORDER BY
  \\"orders1\\".\\"Details\\" ASC
LIMIT 2000
"
`;

exports[`Orders Builder should add foreign column order properly
      and hide already added column 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
  LEFT JOIN \\"products\\" \\"products1\\" ON \\"orders1\\".\\"ProductID\\" = \\"products1\\".\\"ID\\"
  LEFT JOIN \\"currency\\" \\"currency1\\" ON \\"products1\\".\\"CurrencyID\\" = \\"currency1\\".\\"ID\\"
ORDER BY
  \\"orders1\\".\\"Created At\\" ASC,
  \\"currency1\\".\\"name\\" ASC
LIMIT 2000
"
`;

exports[`Orders Builder should remove all orders except the grouping columns
      when a grouping added and show only picked grouping column
      available for order column 1`] = `
"SELECT
  \\"orders1\\".\\"Details\\" AS \\"Details (orders1)\\"
FROM
  \\"orders\\" \\"orders1\\"
GROUP BY
  \\"orders1\\".\\"Details\\"
ORDER BY
  \\"orders1\\".\\"Details\\" DESC
LIMIT 2000
"
`;

exports[`Orders Builder should update order 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
ORDER BY
  \\"orders1\\".\\"Created At\\" ASC
LIMIT 2000
"
`;
